clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -283,10 +281,15 @@ public Long position(TopicPartition tp) {
 
     public Map<TopicPartition, OffsetAndMetadata> allConsumed() {
         Map<TopicPartition, OffsetAndMetadata> allConsumed = new HashMap<>();
-        for (Map.Entry<TopicPartition, TopicPartitionState> entry : assignment.entrySet()) {
+        
+        for (Map.Entry<TopicPartition, TopicPartitionState> entry : this.assignment.entrySet()) {
+            TopicPartition tp = entry.getKey();
             TopicPartitionState state = entry.getValue();
             if (state.hasValidPosition())
-                allConsumed.put(entry.getKey(), new OffsetAndMetadata(state.position));
+                allConsumed.put(tp, new OffsetAndMetadata(state.position));
+            // remove the unsubscribed topic partition from assigned partitions now that its offset is committed
+            if (state.pendingUnsubscribe)
+                this.assignment.remove(tp);
```

Comment:
This mutation is a little surprising. I wonder if we should make removal explicit with a separate method (e.g. `clearPendingUnsubscribe()`?